src/mito2/src/manifest/manager.rs

```diff
@@ -327,68 +346,116 @@ impl RegionManifestManagerInner {
+                            self.manifest.metadata.region_id, action
+                        );
+                    }
+                }
+            }
+            last_version = version;
+            compacted_actions += 1;
+        }
 
-    async fn last_checkpoint(&self) -> Result<Option<RegionCheckpoint>> {
-        todo!()
-    }
+        if compacted_actions == 0 {
+            return Ok(None);
+        }
 
-    // from Checkpoint
+        let region_manifest = manifest_builder.try_build()?;
+        let checkpoint = RegionCheckpoint {
+            last_version,
+            compacted_actions,
+            checkpoint: Some(region_manifest),
+        };
 
-    /// Set a protocol action into checkpoint
-    pub fn set_protocol(&mut self, _action: ProtocolAction) {
-        todo!()
-    }
+        self.store
+            .save_checkpoint(last_version, &checkpoint.encode()?)
+            .await?;
+        // TODO(ruihang): this task can be detached
+        self.store.delete_until(last_version, true).await?;
 
-    /// The last compacted action's version of checkpoint
-    pub fn last_version(&self) -> ManifestVersion {
-        todo!()
+        info!(
+            "Done manifest checkpoint, start_version: {}, last_version: {}, compacted actions: {}",
+            start_version, last_version, compacted_actions
+        );
+        Ok(Some(checkpoint))
     }
 
-    /// Encode this checkpoint into a byte vector
-    pub fn encode(&self) -> Result<Vec<u8>> {
-        todo!()
-    }
+    /// Fetch the last [RegionCheckpoint] from storage.
+    pub(crate) async fn last_checkpoint(
+        store: &ManifestObjectStore,
+    ) -> Result<Option<RegionCheckpoint>> {
+        let last_checkpoint = store.load_last_checkpoint().await?;
 
-    pub fn decode(_bytes: &[u8], _reader_version: ProtocolVersion) -> Result<Self> {
-        todo!()
+        if let Some((version, bytes)) = last_checkpoint {
+            let checkpoint = RegionCheckpoint::decode(&bytes)?;
+            assert!(checkpoint.last_version() >= version);
+            if checkpoint.last_version() > version {
+                // It happens when saving checkpoint successfully, but failed at saving checkpoint metadata(the "__last_checkpoint" file).
+                // Then we try to use the old checkpoint and do the checkpoint next time.
+                // If the old checkpoint was deleted, it's fine that we return the latest checkpoint.
+                // The only side effect is leaving some unused checkpoint files,
+                // and they will be purged by gc task.
+                warn!("The checkpoint manifest version {} in {} is greater than checkpoint metadata version {}.", store.path(), checkpoint.last_version(), version);
+
+                if let Some((_, bytes)) = store.load_checkpoint(version).await? {
+                    let old_checkpoint = RegionCheckpoint::decode(&bytes)?;
+                    return Ok(Some(old_checkpoint));
+                }
```

Comment:
This is impossible since we always read `__last_checkpoint` first.

Reply:
cc @killme2008

Reply:
Sorry, I don't get the point.

Reply:
`version` is always equal to `checkpoint.last_version()`, so the `if` statement is always false and the code block above is unreachable. We should remove it as it is confusing.

Reply:
Got it. Yes, the case will not happen under the current `load_last_checkpoint` implementation. I forgot why I added this check, maybe the `load_last_checkpoint` was changed before.